### ISSUE TITLE
Hotswapping an adapter can zero another adapter with a shared name prefix

### ISSUE THREAD

**Issue by aswanth-07** [huggingface/peft#3780]
### System Info

- PEFT `main`: `50a277e7c87db460ef7444055788f9da29f2da71`
- PEFT: `0.21.1.dev0`
- Transformers: `5.5.4`
- PyTorch: `2.11.0+cu130`
- Python: `3.13.3`
- Platform: Windows 11, CPU reproduction
- The reproduction uses a local `nn.Linear` model and downloads no checkpoint.

### Who can help?

@BenjaminBossan

### Reproduction

Hotswapping an adapter can silently zero another loaded adapter when their names share a prefix. In this example, replacing `foo` also changes `foobar`.

```python
import torch
from torch import nn

from peft import LoraConfig, get_peft_model
from peft.utils.hotswap import hotswap_adapter_from_state_dict


class Tiny(nn.Module):
    def __init__(self):
        super().__init__()
        self.lin = nn.Linear(4, 4, bias=False)

    def forward(self, x):
        return self.lin(x)


torch.manual_seed(0)
config = LoraConfig(target_modules=["lin"], r=2, lora_alpha=2, lora_dropout=0.0)
model = get_peft_model(Tiny(), config, adapter_name="foo")
model.add_adapter("foobar", config)
layer = model.base_model.model.lin

with torch.no_grad():
    layer.lora_A["foobar"].weight.fill_(1.0)
    layer.lora_B["foobar"].weight.fill_(1.0)

model.set_adapter("foobar")
inputs = torch.ones(2, 4)
with torch.no_grad():
    output_before = model(inputs).clone()
    adapter_a_before = layer.lora_A["foobar"].weight.clone()
    adapter_b_before = layer.lora_B["foobar"].weight.clone()

incoming = {
    "base_model.model.lin.lora_A.foo.weight": torch.full_like(layer.lora_A["foo"].weight, 2.0),
    "base_model.model.lin.lora_B.foo.weight": torch.full_like(layer.lora_B["foo"].weight, 3.0),
}
hotswap_adapter_from_state_dict(model, incoming, adapter_name="foo", config=config)

with torch.no_grad():
    output_after = model(inputs)

torch.testing.assert_close(layer.lora_A["foobar"].weight, adapter_a_before)
torch.testing.assert_close(layer.lora_B["foobar"].weight, adapter_b_before)
torch.testing.assert_close(output_after, output_before)
```

The first assertion fails because all eight values in `lora_A["foobar"].weight` change from `1` to `0`. The second adapter's output changes as a result.

The key selection in `hotswap_adapter_from_state_dict` currently checks `adapter_name in k`. Both `foo` and `foobar` therefore enter `missing_keys`. Only the incoming `foo` keys are removed, and the remaining `foobar` keys are zeroed as if they were missing from the replacement checkpoint.

I searched open and closed PEFT issues and pull requests for hotswap adapter-name prefixes, substring matching, unrelated adapter mutation, and zeroed adapters. I found no overlapping report or implementation. The two current hotswap PRs cover dummy adapter configuration and Conv2d dilation and padding, not adapter-key selection.

### Expected behavior

Hotswapping `foo` should update only the exact `foo` adapter. The weights and output of the separately loaded `foobar` adapter should remain unchanged.

Would a focused PR be welcome that selects hotswap keys by an exact adapter path component and adds a public-flow regression proving that a non-target adapter with a shared name prefix keeps both its weights and output? I will wait for explicit approval before changing source code or opening a PR.

AI assistance disclosure: AI-assisted development tools were used during the repository audit, reproduction, duplicate search, and issue drafting. I ran the reproduction against the stated PEFT commit and reviewed the affected code path.

**Comment by gopaljigaur** [comment]
I'd like to work on this, if nobody has picked it up.

Reading `hotswap_adapter_from_state_dict`, the `missing_keys` set is built with a substring test, `(parameter_prefix in k) and (adapter_name in k)` (`src/peft/utils/hotswap.py:472`), so with adapters `foo` and `foobar` every `foobar` key is also counted as a `foo` key. Those keys never appear in the incoming state dict, so they are treated as missing and handled accordingly. I'll confirm with the reproducer that this is the whole story, then:

- match the adapter name as a whole path segment (e.g. `f".{parameter_prefix}.{adapter_name}."` style boundaries) for `missing_keys` and any other name checks in that file
- add a regression test with `foo` / `foobar` asserting `foobar`'s weights and outputs are unchanged after hotswapping `foo`
- run the hotswap tests and `make quality`

Per CONTRIBUTING I'll wait for a maintainer go-ahead before opening the PR. @BenjaminBossan

**Comment by BenjaminBossan** [comment]
Thanks for reporting, I could reproduce the issue. Please open a PR to fix this. Ideally, re-use the logic from normal adapter loading or try to find a common abstraction if it doesn't fit.

/peft-triage approved
